### DISCUSSION TITLE
大佬，请教一个问题：pnpm run serve 是如何通知浏览器安装或更新用户脚本的呢？

### DISCUSSION THREAD

**Discussion by juedui0769** [lisonge/vite-plugin-monkey#74] [Q&A] [answered]
运行 `pnpm run serve` 会打开浏览器，提示，安装或更新；这是在和油猴插件通信吗？
这个功能是如何实现的呢？请大佬教教我，十分感谢！

**Comment by lisonge** [comment, marked as answer]
> 这是在和油猴插件通信吗？

不是

---

> 这个功能是如何实现的呢？

vite-plugin-monkey 只是调用浏览器打开 `http://127.0.0.1:5173/__vite-plugin-monkey.install.user.js?origin=http://127.0.0.1:5173`

浏览器扩展比如 tampermonkey 会自动识别 `*.user.js` 这种 url 然后打开它的安装页面

<https://github.com/lisonge/vite-plugin-monkey/blob/62c4bf21dc845764a4818b7741ec053c7b7b85f5/packages/vite-plugin-monkey/src/node/plugins/server.ts#L215>

**Comment by juedui0769** [comment]
谢谢@lisonge ，还是大佬厉害！

> 浏览器扩展比如 tampermonkey 会自动识别 *.user.js 这种 url 然后打开它的安装页面

之前我真不知道，原来是油猴有识别`*.user.js`的功能
我在这篇 <https://github.com/lisonge/vite-plugin-monkey/discussions/72> 讲原理的帖子中，看到了 `server.ts`了，也打开看了，但是没搞明白；

非常感谢，学习到新知识了！感谢！

**Comment by LinLin00000000** [comment]
大佬nb！我也仿造着大佬的思路使用 [open 库](https://github.com/sindresorhus/open) 实现了自动打开浏览器安装脚本的功能

```JavaScript
function installScript(url) {
    const htmlContent = `<script>location.href = '${url}'; window.close()</script>`

    const base64Content = Buffer.from(htmlContent).toString('base64')

    const dataUrl = `data:text/html;base64,${base64Content}`

    return open(dataUrl, {
        app: { name: apps.browser },
    })
}
```
